### DISCUSSION TITLE
Repetitive column values optimisation

### DISCUSSION THREAD

**Discussion by ComputMagnac** [mikro-orm/mikro-orm#5899] [Ideas]
**Is your feature request related to a problem? Please describe.**
When using string with json types, it would be great to have the actual column null if the column text is equal to the default value.
This would avoid filling a big JSON object with always the same default values into each column, wich is taking space & processing time

**Describe the solution you'd like**
Examples of a fast custom JSON database system I created below

**Describe alternatives you've considered**
Not really possible in a clean way

Example from a custom homemade JSON database from a specific object, but as you will see on the INSERT, it was also handling any types such as string, booleans

[[SELECT]]
```js
for (let keyName of Object.keys(this.#structure))
{		
    // If the column key is undefined (not in the json) and that the structure column key has a default value
    if (finalObj[keyName] === undefined && this.#structure[keyName].default !== undefined)
        // the selected object key will become the default value available on the structure
	finalObj[keyName] = JSON.parse(JSON.stringify(this.#structure[keyName].default));
}
return finalObj; // To the user
```
[[UPDATE]]
```js
// if the key structure default is an object and the structure key default is the same as the update key value
if (trueTypeOf(this.#structure[keyName].default) === "object" && isObjectFullMatch(this.#structure[keyName].default, values[keyName]))
    // We delete from the json so it will not be inserted into the JSON file (the select will be the one to place it)
    delete this.#data[matchingIndex][keyName];

```

[[INSERT]]
```js
if (this.#structure[keyName].default !== null)
{
    if (this.#structure[keyName].type === "object" && isObjectFullMatch(this.#structure[keyName].default, value))
	continue;
    else if (value === this.#structure[keyName].default)
	continue;

    finalEntry[keyName] = value;
}
this.#data.push(finalEntry);
```

**Comment by B4nan** [comment]
So you would like to ignore values from database, mapping them to `null` if the value matches the db default, do I get that right? You could achieve this behavior with a [custom type](https://mikro-orm.io/docs/custom-types). You could leverage the `convertToJSValueSQL` method to alter the SQL queries to use something like `if` or `case` to map the column to `null` when it matches some value (not sure if you can refer to the db default via SQL statement in such context, normally `default` keyword works like this for `insert/update` queries, worst case you could put the default value there explicitly).

Something like this could do the job (not tested)

```ts
import { JsonType } from '@mikro-orm/core';

export class MyJsonType extends JsonType {

  override convertToDatabaseValueSQL(key: string, platform: Platform): string {
    // the json type already uses this method to add casting in postgres, so you should use the parent method via `super` first
    const sql = super.convertToDatabaseValueSQL(key, platform);
    return `case when ${sql} = ${platform.quoteValue(this.prop.default)} then null else ${sql} end`;
  }

}
```

**Reply by B4nan** [reply]
Or maybe it would be better to keep your actual property as `lazy: true` so you never select it unless you want to, and have a second `formula` property with this SQL expression, only for reading purposes.

```ts
@Property({ type: 'json', default: '...something big...', lazy: true })
myBigJsonColumn: any;

@Property({ formula: alias => `case when ${alias}.my_big_json_column = '...something big...' then null else ${alias}.my_big_json_column end` })
myBigJsonColumn: any;
```

**Reply by ComputMagnac** [reply]
Your first idea looks better (as well as more readable in my opinion), the only downside is that it is not native feature north will it be documented for future uses case.

Also it appears that his will only work for insert, but once we want to select, we would have a null.
Obviously that requires to implement witch is probably not that big of a deal, I just believe it would have been nice to have this as a default feature since it can only benefit the users and I cannot see any downside of implementing this OOTB.

It all depends of the use cases, I think it can only be benefic especially on large amount of dataset, where a ton of storage can be obtained.

One downside is for any user obtaining the sqlite and trying to read it out of Mikro Orm, they won't be able to obtain all thoses columns because they will be all null, but I suppose this is kind of an edge case.

**Reply by B4nan** [reply]
Custom types as well as formulas are a native feature. I don't think what you want is common, so having some magical option just for you feels weird to be honest :] 

Also, if I would be implementing this "natively", I would use the very same approach I described above, there is no need to alter internals here.

> Also it appears that his will only work for insert, but once we want to select, we would have a null.

Custom types are respected in all query types. Maybe this should be `convertToJSValueSQL` instead, or both methods might need to be adjusted. One handles reading data, the other is for writing data.
